# STATISTICAL TESTS Task

In [1]:
import numpy as np
from scipy import stats
from statsmodels.stats import proportion

# import scipy

In [ ]:
def generate_sample(mean, sd, n, seed=None):
    rng = np.random.default_rng(seed)

    arr = rng.normal(size=n)

    # Standardize
    arr = (arr - arr.mean()) / arr.std(ddof=1)

    # Rescale
    arr = arr * sd + mean

    return arr

## Question 1
#### The historical average score in a certification exam is 65. A training institute introduces a new course. A random sample of 50 students trained under the new course has a mean score of 68 and a standard deviation of 10. At the 5% level of significance, test whether the new course improves performance.

1. Noted Points: 
    - Population Mean = 65
    - Sample Size = 50
    - Sample Mean = 68
    - Sample SD = 10
    - Significance Level = 5% = 0.05
2. Hypothesis
    - H0 : mean = 65
    - Ha : mean > 65
3. Test : One Sample t-test Right Tailed


In [3]:
student_score = generate_sample(mean=68, sd=10, n=50, seed=42)

t_stats, p_value = stats.ttest_1samp(a=student_score, popmean=65, alternative="greater")

print(f"t-Statistics: {t_stats}")
print(f"p-Value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do no Reject H0")

t-Statistics: 2.1213203435596526
p-Value: 0.019489027622174224
Reject H0.


1. Results:
   - t-statistic = 2.1213
   - p-value = 0.0194

1. Decision: 
    - p-value (0.0194) < significant level (0.05), reject H0

2. Conclusion:
    - At 5% level of significance level, we reject the null Hypothesis. There is sufficient evidence to conclude that the new course improves student peformance.

----

## Question 2
#### A company claims that the average time taken to resolve customer tickets is 30 minutes. A random sample of 40 tickets shows a mean resolution time of 33 minutes with a standard deviation of 8 minutes. At the 5% level of significance, test whether the average resolution time has increased.



1. Noted Points: 
    - Population Mean = 30
    - Sample Size = 40
    - Sample Mean = 33
    - Sample SD = 8
    - Significance Level = 5% = 0.05
2. Hypothesis
    - H0 : mean = 30
    - Ha : mean > 30
3. Test : One Sample t-test Right Tailed

In [4]:
ticket_resolve_time = generate_sample(mean=33, sd=8, n=40, seed=42)

t_stats, p_value = stats.ttest_1samp(a=ticket_resolve_time, popmean=30, alternative="greater")

print(f"t-Statistics: {t_stats}")
print(f"p-Value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do no Reject H0")

t-Statistics: 2.3717082451262845
p-Value: 0.011369083410373513
Reject H0.


1. Results:
   - t-statistic = 2.3717
   - p-value = 0.0113

1. Decision: 
    - p-value (0.0113) < significant level (0.05), reject H0

2. Conclusion:
    - At 5% level of significance level, we reject the null Hypothesis. There is sufficient evidence to conclude that the average ticket resolution time has increased beyond 30 minutes.

---

## Question 3
#### A company claims that 60% of its users are satisfied with its product. In a survey of 200 users, 134 users report satisfaction. At the 5% level of significance, test whether customer satisfaction is higher than claimed.

In [5]:
134/200

0.67

1. Noted Points: 
    - Population proportion = 60% = 0.60
    - Sample Size = 200
    - Satisfied users = 134
    - Sample proportion = 134/200 = 0.67
    - Significance Level = 5% = 0.05
2. Hypothesis
    - H0 : proportion = 0.60
    - Ha : proportion > 0.60
3. Test : One Sample Proportion z-test Right Tailed

In [6]:
z_stat, p_value = proportion.proportions_ztest(count=134, nobs=200, value=0.60, alternative="larger")

print(f"z-Statistics: {z_stat}")
print(f"p-Value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do no Reject H0")

z-Statistics: 2.105322668110795
p-Value: 0.01763161510876105
Reject H0.


1. Result:
    - z-statics = 2.1053
    - p-value = 0.0176

2. Decision:
    - p-value (0.0176) < Significance Level (0.05), reject H0

3. Conclusion:
    - At 5% significace leve, there is sufficient evidence to conclude that customer satisfication is higher than the claimed 60%.

---

## Question 4
#### Historically, 40% of applicants pass an entrance test. After a new preparation program, 90 out of 180 applicants pass. At the 5% level of significance, test whether the pass rate has improved.

1. Noted Points: 
    - Population proportion = 40% = 0.40
    - Sample Size = 180
    - Satisfied users = 90
    - Sample proportion = 90/180 = 0.50
    - Significance Level = 5% = 0.05
2. Hypothesis
    - H0 : proportion = 0.40
    - Ha : proportion > 0.40
3. Test : One Sample Proportion z-test Right Tailed

In [7]:
z_stat, p_value = proportion.proportions_ztest(count=90, nobs=180, value=0.40, alternative="larger")

print(f"z-Statistics: {z_stat}")
print(f"p-Value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do no Reject H0")

z-Statistics: 2.683281572999747
p-Value: 0.0036451790457678298
Reject H0.


1. Result:
    - z-statics = 2.6832
    - p-value = 0.00364

2. Decision:
    - p-value (0.00364) < Significance Level (0.05), reject H0

3. Conclusion:
    - At 5% significace leve, there is sufficient evidence to conclude that new preparation program has pass rate higher than 40%

---

## Question 5
#### Two teaching methods are used to train employees. A sample of 30 employees trained using Method A has a mean score of 75 with a standard deviation of 12. A sample of 28 employees trained using Method B has a mean score of 70 with a standard deviation of 10. At the 5% level of significance, test whether there is a difference in average scores between the two methods.



1. Noted Points:
    - Method A:
        - Sample Size = 30
        - Sample Mean = 75
        - Sample SD = 12
    - Method B:
        - Sample Size = 28
        - Sample Mean = 70
        - Sample SD = 10
    - Significance leve = 5% = 0.05

2. Hypothesis:
    - H0 = mean(A) = mean(B)
    - Ha = mean(A) != mean(B)

3. Test:
    - Two-Sample t-test Two Tailed ( Welch's t-test)

In [8]:
method_a = generate_sample(mean=75, sd=12, n=30, seed=42)
method_b = generate_sample(mean=70, sd=10, n=28, seed=123)

t_stats, p_value = stats.ttest_ind(
    a=method_a, b=method_b, equal_var=False, alternative="two-sided"
)

print(f"t-Statistics: {t_stats}")
print(f"p-Value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do no Reject H0")

t-Statistics: 1.7281053564140343
p-Value: 0.08954847598333188
Do no Reject H0


1. Result:
    - t-statistics = 1.7281
    - p-value = 0.0895

2. Decision:
    - p-value (0.0895) > Significance Level (0.05), do not reject H0

3. Conclusion:
    - At the 5% significance level, we do not reject null hypothesis. There is insufficient evidence to conclude that the average scores differ between method A and method B. 

---

## Question 6
#### The average delivery time from two warehouses is compared using random samples. Warehouse A has a sample of 35 deliveries with a mean time of 48 minutes and a standard deviation of 6 minutes. Warehouse B has a sample of 40 deliveries with a mean time of 52 minutes and a standard deviation of 7 minutes. At the 5% level of significance, test whether the mean delivery times differ.



1. Noted Points:
    - Warhouse A:
        - Sample Size = 35
        - Sample Mean = 48
        - Sample SD = 6
    - Warhous B:
        - Sample Size = 40
        - Sample Mean = 52
        - Sample SD = 7
    - Significance leve = 5% = 0.05

2. Hypothesis:
    - H0 = mean(A) = mean(B)
    - Ha = mean(A) != mean(B)

3. Test:
    - Two-Sample t-test Two Tailed ( Welch's t-test)

In [9]:
warhouse_a = generate_sample(mean=48, sd=6, n=35, seed=42)
warhouse_b = generate_sample(mean=52, sd=7, n=40, seed=123)

t_stats, p_value = stats.ttest_ind(
    a=warhouse_a, b=warhouse_b, equal_var=False, alternative="two-sided"
)

print(f"t-Statistics: {t_stats}")
print(f"p-Value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do no Reject H0")

t-Statistics: -2.6645527807485094
p-Value: 0.009483418770892945
Reject H0.


1. Result:
    - t-statistics = -2.6645
    - p-value =  0.0094

2. Decision:
    - p-value ( 0.0094) < Significance Level (0.05), reject H0

3. Conclusion:
    - At the 5% significance level,we reject null hypothesis. There is sufficient evidence to conclude that the average delivery time differ between Warhouse A and Warhouse B

---

## Question 7
#### Two versions of a website are tested. Out of 500 users for Version A, 320 users click the signup button. Out of 480 users for Version B, 290 users click the signup button. At the 5% level of significance, test whether the conversion rates differ.

In [10]:
320/500, 290/480

(0.64, 0.6041666666666666)

1. Noted Points:
    - Verion A:
        - Sample Size = 500
        - Signups = 320
        - Sample Proportion = 320/500 = 0.64
    - Vesion B:
        - Sample Size = 480
        - Signups = 290
        - Sample Proportion = 290/480 = 0.6042
    - Significance leve = 5% = 0.05

2. Hypothesis:
    - H0 = proportion(A) = proportion(B)
    - Ha = proportion(A) != proportion(B)

3. Test:
    - Two-Sample proportional z-test Two Tailed

In [11]:
z_stat, p_value = proportion.proportions_ztest(
    count=[320, 290], nobs=[500, 480], alternative="two-sided"
)

print(f"z-Statistic: {z_stat}")
print(f"p-Value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do Not Reject H0.")

z-Statistic: 1.1567515462334546
p-Value: 0.24737388327470267
Do Not Reject H0.


1. Result:
    - z-statistic = 1.1567
    - p-value = 0.2473

2. Decision:
    - p-value(0.2473) > significant level (0.05), do not reject H0

3. Conclustion
    - At 5% significance leavel, we do not reject the null hypothesis. There is insufficient evidence to conclude that the conversion reates differ between version A and version B.

---

## Question 8
#### A vaccine is tested in two regions. In Region A, 180 out of 300 people recover. In Region B, 165 out of 310 people recover. At the 5% level of significance, test whether the recovery rates are significantly different.

In [12]:
180 / 300, 165 / 310

(0.6, 0.532258064516129)

1. Noted Points:
    - Region A:
        - Sample Size = 300
        - recovered = 180
        - Sample Proportion = 180/300 = 0.6
    - Region B:
        - Sample Size = 310
        - recovered = 165
        - Sample Proportion = 165/310 = 0.5323
    - Significance leve = 5% = 0.05

2. Hypothesis:
    - H0 = proportion(A) = proportion(B)
    - Ha = proportion(A) != proportion(B)

3. Test:
    - Two-Sample proportional z-test Two Tailed

In [13]:
z_stat, p_value = proportion.proportions_ztest(
    count=[180, 165], nobs=[300, 310], alternative="two-sided"
)

print(f"z-Statistic: {z_stat}")
print(f"p-Value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do Not Reject H0.")

z-Statistic: 1.6874523112864297
p-Value: 0.09151641239095082
Do Not Reject H0.


1. Result:
    - z-statistic = 1.6874
    - p-value = 0.0915

2. Decision:
    - p-value(0.0915) > significant level (0.05), do not reject H0

3. Conclustion
    - At 5% significance leavel, we do not reject the null hypothesis. There is insufficient evidence to conclude that the recovery rates are different between region A and region B.

---

## Question 9
#### A company wants to compare the average productivity of employees trained using three different training programs. Program A has 25 employees with a mean productivity score of 80 and a standard deviation of 8. Program B has 30 employees with a mean of 75 and a standard deviation of 7. Program C has 28 employees with a mean of 82 and a standard deviation of 6. At the 5% level of significance, test whether there is a difference in mean productivity across the programs

1. Noted Points:
    - Program A:
        - Sample Size = 25
        - Sample Mean = 80
        - Sample SD = 8
    - Program B:
        - Sample Size = 30
        - Sample Mean = 75
        - Sample SD = 7
    - Program C:
        - Sample Size = 28
        - Sample Mean = 82
        - Sample SD = 6
    - Significance leve = 5% = 0.05

2. Hypothesis:
    - H0 = mean(A) = mean(B) = mean(B)
    - Ha = At least one program mean differs

3. Test:
    - One-Way ANOVA (Welch's One-Way ANOVA)

In [14]:
program_a = generate_sample(mean=80, sd=8, n=25, seed=42)
program_b = generate_sample(mean=75, sd=7, n=30, seed=123)
program_c = generate_sample(mean=82, sd=6, n=28, seed=456)


f_stats, p_value = stats.f_oneway(program_a, program_b, program_c, equal_var=False)

print(f"f-statistic: {f_stats}")
print(f"p-value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do Not Reject H0.")

f-statistic: 8.462063147550968
p-value: 0.0006668560021552565
Reject H0.


1. Result:
    - f-statistic = 8.4620
    - p-value = 0.0006

2. Decision:
    - p-value(0.0006) < significant level (0.05), reject H0

3. Conclustion
    - At 5% significance leavel, we reject the null hypothesis. There is sufficient evidence to conclude that at least one training program has a differnct mean productivity scroe.

---

## Question 10
#### 10.	A company wants to test whether gender and product preference are independent. The survey results are as follows:
- #### Male: 60 prefer Product A, 40 prefer Product B
- #### Female: 50 prefer Product A, 50 prefer Product B 
#### At the 5% level of significance, test for independence.

1. Noted Points
    - Male:
        - Product A = 60
        - product B = 40
    - Femal:
        - Product A = 50
        - product B = 50
    - significance level = 5% = 0.05

2. Hypothesis
    - H0: Gender and Product preference are independent
    - Ha: Gender and Product preference are not independent

3. Test: Chi- Square Test of Independence

In [15]:
table = [
    [60, 40],
    [50, 50]
]

chi2, p_value, dof, expected = stats.chi2_contingency(observed=table)


print(f"chi-square Statistic: {chi2}")
print(f"p-value: {p_value}")
print(f"Expected: \n{expected}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do Not Reject H0.")

chi-square Statistic: 1.6363636363636362
p-value: 0.20082512269514174
Expected: 
[[55. 45.]
 [55. 45.]]
Do Not Reject H0.


1. Result:
    - Chi-square Statistic = 1.6363
    - p-value = 0.2008

2. Decision:
    - p_value (0.2008) > significant level (0.05), do not reject Null Hypothesis

3. At 5% level of significance, we do not reject null hypothesis. There is insufficient evidence to conclude that gender and product preference are associated. Therefore, the data suggest that gendera and product preference are independent.

---

## Question 11
#### A  hospital studies whether smoking status is related to disease occurrence. The data collected is as follows:
- #### Smokers: 70 have the disease, 30 do not
- #### Non-smokers: 40 have the disease, 60 do not
#### At the 5% level of significance, test whether smoking status and disease occurrence are independent. 


1. Noted Points
    - Smokers:
        - Disease = 70
        - No Disease = 30
    - Non-Smokers:
        - Disease = 40
        - No Disease = 60
    - significance level = 5% = 0.05

2. Hypothesis
    - H0: Smoking status and Disease occurance are independent
    - Ha: Smoking status and Disease occurance are not independent

3. Test: Chi- Square Test of Independence

In [16]:
table = [
    [70, 30],
    [40, 60]
]

chi2, p_value, dof, expected = stats.chi2_contingency(table)

print(f"Chi-square Statistic: {chi2}")
print(f"p-value: {p_value}")
print(f"Expected:\n{expected}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do Not Reject H0.")

Chi-square Statistic: 16.98989898989899
p-value: 3.7579211012441855e-05
Expected:
[[55. 45.]
 [55. 45.]]
Reject H0.


1. Result:
    - Chi-square Statistic = 16.9898
    - p-value = 3.7579211012441855e-05 = 0.00003757

2. Decision:
    - p_value (0.00003757) < significant level (0.05), reject Null Hypothesis

3. At 5% level of significance, we reject null hypothesis. There is sufficient evidence to conclude that Smoking status and Disease occurance are associated. Therefore, the Smoking status and Disease occurance are not independent

---

## Question 12
#### A dice manufacturer claims that a die is fair. The die is rolled 120 times, and the observed frequencies are:
- #### Face 1: 18
- #### Face 2: 22
- #### Face 3: 20
- #### Face 4: 25
- #### Face 5: 17
- #### Face 6: 18
#### At the 5% level of significance, test whether the die is fair.

In [17]:
# Expected frequency for each face:
120/6

20.0

1. Noted Points:
    - Observed Frequencies:
        - Face 1 = 18
        - Face 2 = 22
        - Face 3 = 20
        - Face 4 = 25
        - Face 5 = 17
        - Face 6 = 18
    - Total Rolls = 120
    - Expected Frequenies = [20, 20, 20, 20, 20, 20]
    - Significance Level = 5% = 0.05

2. Hypothesis
    - H0: The die is fair
    - Ha: The die is not fair
3. Test: Chi-Square Goodness-of-Fit Test

In [18]:
observed = [18, 22, 20, 25, 17, 18]
expected = [20] * 6

chi2, p_value  = stats.chisquare(f_obs=observed, f_exp=expected)

print(f"Chi-square Statistic: {chi2}")
print(f"P-value: {p_value}")

if p_value < 0.05:
    print("Reject H0.")
else:
    print("Do Not Reject H0.")

Chi-square Statistic: 2.3000000000000003
P-value: 0.8062668698851285
Do Not Reject H0.


1. Result:
    - Chi-square Statistic = 2.30
    - p-value ≈ 0.807

2. Decision:
    - p-value (0.807) > significant level (0.05), do not reject H0.

3. Conclusion:
    - At the 5% significance level, we do not reject the null hypothesis. There is insufficient evidence to conclude that the die is unfair. The observed frequencies are consistent with a fair die.

---